### PR TITLE
clang: detect host Android API level

### LINKED ISSUE #2469: Clang should specify __ANDROID_API__ to current device version
Some C functions require `__ANDROID_API__` to be set.

For example, the following will not compile without 'implicit declaration' errors on Nougat+:
```c
#define _GNU_SOURCE
#include <stdio.h>
#include <string.h>

int main(void) {
    const char *hello_world = "Hello, world!";
    char *world = strchrnul(hello_world, 'w');
    if (world)
        puts(world);
    return 0;
}
```
```
$ clang test.c
test.c:7:16: warning: implicit declaration of function 'strchrnul' is invalid in      C99 [-Wimplicit-function-declaration]
        char *world = strchrnul(hello_world, 'w');
                      ^
test.c:7:8: warning: incompatible integer to pointer conversion initializing
      'char *' with an expression of type 'int' [-Wint-conversion]
        char *world = strchrnul(hello_world, 'w');
              ^       ~~~~~~~~~~~~~~~~~~~~~~~~~~~
2 warnings generated.
```
Compile with `__ANDROID_API__` defined to 24 or higher (assuming, again, that you are running Nougat or above), and it works fine.

Clang does have a config file that can do that, but in order for the file to be loaded, you would have to set the CMake flag `-DCLANG_CONFIG_FILE_SYSTEM_DIR` while compiling Clang to a valid path, then it will automatically load the config file.

After that, all you would need to do is add a shell script to the postinstall script that does something like this:
```sh
#!/data/data/com.termux/files/usr/bin/dash
API_STR="-D__ANDROID_API__="
API_LEVEL="$(getprop ro.build.version.sdk)"
CONFIG_FILE="<config file path>"

echo "Updating Clang's config file..."

# Check if the file exists and it contains "-D__ANDROID_API__="
if [ -f "${CONFIG_FILE}" ] && [ "$(busybox grep -qs "${API_STR}" "${CONFIG_FILE}")" ]
then
    # Check if this config file is from an older Android version, if so, update it
    if [ ! "$(busybox grep -qs "${API_STR}${API_LEVEL}" "${CONFIG_FILE}")" ]
    then
        busybox sed -i'' -e "s#${API_STR}[0-9]\*#${API_STR}${API_LEVEL}#g" "${CONFIG_FILE}"
    fi
    # Otherwise, do nothing as the file has the latest version
else
    # We now can just append __ANDROID_API__ to the file.
    echo "${API_STR}${API_LEVEL}" >> "${CONFIG_FILE}"
fi
```
Make sure it gets tested, I just wrote this blindly. I am not the greatest at shell scripts or `sed`, and I don't know much about dpkg either.

If this works properly, Clang should automatically define `__ANDROID_API__`, and that source file should compile without issues, again, assuming you are running Nougat or later.

### PR DESCRIPTION
**Pull request by truboxl**
Fixes #2469

### CODE REVIEW

File: packages/libllvm/build.sh [hunk 1]
```diff
@@ -88,9 +88,13 @@ termux_step_pre_configure() {
...
 
-	# Add unknown vendor, otherwise it screws with the default LLVM triple
-	# detection.
+	# Change first hyphen to add unknown vendor,
+	# otherwise it screws with the default LLVM triple detection.
+	# Remove default Termux Android API level in LLVM triple and
+	# handle detection in clang.
 	export LLVM_DEFAULT_TARGET_TRIPLE=${CCTERMUX_HOST_PLATFORM/-/-unknown-}
+	LLVM_DEFAULT_TARGET_TRIPLE=${LLVM_DEFAULT_TARGET_TRIPLE/androideabi${TERMUX_PKG_API_LEVEL}/androideabi}
+	LLVM_DEFAULT_TARGET_TRIPLE=${LLVM_DEFAULT_TARGET_TRIPLE/android${TERMUX_PKG_API_LEVEL}/android}
```
Comment (buttaface) [inline packages/libllvm/build.sh:97]: I don't think we should change this default, as it is also used to compile LLVM for Android.

**Comment by buttaface** [comment]
I'm skeptical of this change, as we have a long precedent of building for API 24 now, even though we always linked against the system libc, which usually had later APIs. Since you can always override the default by manually specifying the triple you want, I think that is enough.

I'm leaning against this in principle, @Grimler91, @sylirre, @xtkoba, let me know what you think.

File: packages/libllvm/build.sh (hunk 1 above)
Comment (truboxl) [reply packages/libllvm/build.sh:97]: This changes default target in `clang -v` and defines in `echo | clang -E -dM -` which omitted the API level.

For building LLVM will rely on `<android/api-level.h>` instead which is not touched at all and kept at API level 24.

Comment (buttaface) [reply packages/libllvm/build.sh:97]: > For building LLVM, we already put --target=arch-unknown-linux-android24 everywhere in build script

Right, but [we're also setting `LLVM_HOST_TRIPLE` to this just below](https://github.com/termux/termux-packages/blob/e20c8974b4acc2c6ffb65b235595dd61d1d403bc/packages/libllvm/build.sh#L110), so if our `CFLAGS/CXXFLAGS` are not pulled in properly somewhere, we'd be building that LLVM file with a target triple without an API level.

I would prefer to keep this the same when building clang for Android, and only override it when running clang _on_ Android.

Anyway, not asking you to change this now, as I'm leaning against this whole pull atm.

**Comment by xtkoba** [comment]
Backward compatibility is not important here. It should be enough to announce that the default is changed, if ever necessary. Unlikely to be anything catastrophic.

Assume that we provide two types of commands. One is for the minimum API level that Termux supports (24 as of now) and the other is for the device API level. Then what should be the target prefix? To me it is natural that `aarch64-linux-android24-` is assigned to the former and `aarch64-linux-android-` (without API number) to the latter. But it is acceptable that `aarch64-linux-android-` is assigned to the minimum API level if another appropriate prefix is given to the device API level. Maybe `native-` (`native-clang`) or something. For user's convenience, it is preferable that we provide such commands that target to the device API level, if not the default.

**Comment by buttaface** [comment]
> But it is acceptable that aarch64-linux-android- is assigned to the minimum API level if another appropriate prefix is given to the device API level. Maybe native- (native-clang) or something. For user's convenience, it is preferable that we provide such commands that target to the device API level, if not the default.

That would be very easy to do and wouldn't require these build changes: just add a `native-clang` shell script to the clang package that similarly gets the device API level and runs `clang --target=$arch-unknown-linux-android$device_api_level` instead. I'd be fine with such an addition, rather than changing the default after so many years.

**Comment by xtkoba** [comment]
And someone can provide an unofficial package that replaces `clang` with `native-clang`. Everyone will find and use it if well advertised. I wonder why it is not done yet.

**Comment by truboxl** [comment]
Sure shell scripts will work too. But complexity will be in packaging them in a non intrusive way that doesn't interfere with the existing symlinks. I expect some symlinks will be replaced if follow this route.

Can close this if alternative is provided.

File: packages/libllvm/clang-lib-Basic-Targets-OSTargets.h.patch [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+ namespace targets {
+ 
+@@ -379,6 +389,14 @@
+       this->PlatformName = "android";
+       this->PlatformMinVersion = Triple.getEnvironmentVersion();
+       const unsigned Maj = this->PlatformMinVersion.getMajor();
++      char host_api_version[PROP_VALUE_MAX];
++      const int len = __system_property_get("ro.build.version.sdk", host_api_version);
++      if (len < 1) {
++        memcpy(host_api_version, TERMUX_PKG_API_LEVEL, sizeof(host_api_version));
```
Comment (licy183) [inline packages/libllvm/clang-lib-Basic-Targets-OSTargets.h.patch:31]: This should be `strcpy(host_api_version, TERMUX_PKG_API_LEVEL)`. Using `memcpy` may lead to intended behavior.

**Review by licy183 — CHANGES_REQUESTED** [review]

File: packages/libllvm/clang-lib-Basic-Targets-OSTargets.h.patch (hunk 2 above)
Comment (landfillbaby) [reply packages/libllvm/clang-lib-Basic-Targets-OSTargets.h.patch]: resolved by using `strncpy`

File: packages/libllvm/build.sh (hunk 1 above)
Comment (truboxl) [reply packages/libllvm/build.sh:97]: When API level is not specified, build will fallback to use `<android/api-level.h>`

**Comment by buttaface** [comment]
@truboxl, I suggest you submit a script named something like `device-api-clang` instead, and I'd be fine with getting that in. Otherwise, this is going to move a bunch of users who were compiling on-device fine with the default of API 24 to building against the device API instead, potentially breaking working builds. I am not worried about it for my own code, but I suspect others will complain.

I'm not completely against this change though, which is why I'm waiting for other Termux devs to chime in with their opinion. It is an open question of whether building against the minimum API 24 by default is better or the higher device API, and if the benefit of shifting the default API built is worth the cost of breaking builds.

**Comment by truboxl** [comment]
Agreed more discussion is needed. That #2469 is long overdue with many unhappy users in there.

It can be fixed either by source (as demonstrated by this PR) or wrappers (will check on this). I am also open to other alternatives with the main goal of running just `clang`, `clang++` and others can autodetect API level without specifying extra argument.

**Comment by landfillbaby** [comment]
> wrappers (will check on this).
```sh
#!@TERMUX_PREFIX@/usr/bin/sh
clang --target="arch-unknown-linux-android$(getprop ro.build.version.sdk)"
```
and a clang++ version ofc

**Comment by truboxl** [comment]
Abandoned. Need to accommodate adding extra api levels to `ndk-multilib-*` if this goes through which I probably won't have time for that. I guess people are more comfortable back porting to Android 7 now...

**Pull request closed without merging**